.github/workflows/image-builder.yml

```diff
@@ -0,0 +1,112 @@
+on:
+  workflow_call:
+    inputs:
+      group:
+        description: 'A name for the group of images to build'
+        required: true
+        type: string
+      base-images:
+        description: 'The base images to build'
+        required: false
+        type: string
+        default: ''
+      images:
+        description: 'The images to build'
+        required: true
+        type: string
+
+env:
+  COMMIT_SHA: ${{ github.event.pull_request.head.sha || github.sha }}
+  # Secrets aren't accessible for PRs coming from forks, and we also can't
+  # check secrets in an `if` conditional, so we check for the presence of
+  # the secret here and use this value in conditionals below.
+  CAN_PUSH_IMAGE: ${{ secrets.DOCKERHUB_USERNAME != '' }}
+
+jobs:
+  build-images:
+    strategy:
+      fail-fast: false
+      matrix:
+        os: [ubuntu-24.04, ubuntu-24.04-arm]
+
+    runs-on: ${{ matrix.os }}
+
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          # We need the whole history so we can diff against `master` to determine
+          # what images need to be built.
+          fetch-depth: 0
+
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v3 # https://github.com/marketplace/actions/docker-setup-buildx
+        with:
+          # This is needed to allow pushing to the local registry running on port 5000.
+          driver-opts: network=host
+      - name: Login to DockerHub
+        if: ${{ env.CAN_PUSH_IMAGE == 'true' }}
+        uses: docker/login-action@v3 # https://github.com/marketplace/actions/docker-login
+        with:
+          username: ${{ secrets.DOCKERHUB_USERNAME }}
+          password: ${{ secrets.DOCKERHUB_TOKEN }}
+
+      - name: Create directory for metadata
+        run: mkdir -p ${{ runner.temp }}/image-metadata-${{ inputs.group }}
+
+      - name: Build images
+        run: python3 scripts/build_images.py
+        env:
+          BASE_IMAGES: ${{ inputs.base-images }}
+          IMAGES: ${{ inputs.images }}
+          TAG: ${{ env.COMMIT_SHA }}
+          PUSH_IMAGES: ${{ env.CAN_PUSH_IMAGE == 'true' }}
+          METADATA_DIR: ${{ runner.temp }}/image-metadata-${{ inputs.group }}
```

Comment:
I didn't want to deal with argument parsing, so I'm (ab)using environment variables. This reads a little more closely to invocation of a GitHub action, which I actually sort of like.